frontend/views/containers/chatroom/file-attachment/ChatAttachmentPreview.vue

```diff
@@ -1,46 +1,162 @@
+              @click='deleteAttachment(entry)'
+            )
+              i.icon-trash-alt
+
+  template(v-else)
+    .c-attachment-preview(
+      v-for='entry in attachmentList'
+      :key='entry.attachmentId'
+      :class='"is-" + entry.attachType'
     )
-      i.icon-times
+      img.c-preview-img(
+        v-if='entry.attachType === "image" && entry.url'
+        :src='entry.url'
+        :alt='entry.name'
+      )
+      .c-preview-non-image(v-else)
+        .c-non-image-icon
+          i.icon-file
+
+        .c-non-image-file-info
+          .c-file-name.has-ellipsis {{ entry.name }}
+          .c-file-ext {{ fileExt(entry) }}
+
+      button.c-attachment-remove-btn(
+        type='button'
+        :aria-label='L("Remove attachment")'
+        @click='$emit("remove", entry.url)'
+      )
+        i.icon-times
+
+      .c-loader(v-if='isForDownload && !entry.downloadData')
+
+  a.c-invisible-link(ref='downloadHelper')
 </template>
 
 <script>
+import sbp from '@sbp/sbp'
+import Tooltip from '@components/Tooltip.vue'
+import { ATTACHMENT_TYPES } from '@model/contracts/shared/constants.js'
+
 export default {
   name: 'ChatAttachmentPreview',
+  components: {
+    Tooltip
+  },
   props: {
     attachmentList: {
       // [ { url: string, name: string, attachType: enum of ['image', 'non-image'] }, ... ]
       type: Array,
       required: true
+    },
+    isForDownload: {
+      type: Boolean,
+      default: false
+    }
+  },
+  data () {
+    return {
+      isPreparingDownload: [],
+      preloadedBlobs: {}
+    }
+  },
+  computed: {
+    shouldPreviewImages () {
+      return !this.attachmentList.some(attachment => attachment.attachType === ATTACHMENT_TYPES.NON_IMAGE)
+    }
+  },
+  mounted () {
+    if (this.shouldPreviewImages) {
+      (async () => {
+        for await (const attachment of this.attachmentList) {
+          const blobWithURL = await this.getBlobWithURLFromAttachment(attachment)
+          this.preloadedBlobs[attachment.attachmentId] = blobWithURL
+        }
+        this.$forceUpdate()
+      })()
     }
   },
   methods: {
     fileExt ({ name }) {
       const lastDotIndex = name.lastIndexOf('.')
       const ext = lastDotIndex === -1 ? '' : name.substring(lastDotIndex + 1)
       return ext.toUpperCase()
+    },
+    deleteAttachment (attachment) {
+      console.log('TODO - delete attachment')
+    },
+    async getBlobWithURLFromAttachment (attachment) {
+      if (!attachment.downloadData) { return }
+      const blob = await sbp('chelonia/fileDownload', attachment.downloadData)
+      const url = URL.createObjectURL(blob)
+      return { blob, url }
+    },
+    async downloadAttachment (attachment) {
+      if (!attachment.downloadData) { return }
+
+      // reference: https://blog.logrocket.com/programmatically-downloading-files-browser/
+      const { name, attachmentId } = attachment
+      this.isPreparingDownload = [...this.isPreparingDownload, attachmentId]
```

Comment:
The initial intention of me creating `this.isPreparingDownload` array here is for below two possible uses

- To give visual indicator(blurring or darkening etc..) that  the attachment item that user has just clicked on is being prepared for downloading
- To prevent from calling `sbp('chelonia/fileDownload')` multiple times on the same item while the blob is loading from the server.

If you think this is not needed in your work on graceful handling of large size message, pls go ahead and removing this list.